text/0000-glimmer-components.md

```diff
@@ -0,0 +1,835 @@
+component has been fully created. It is closest to `didInsertElement` in legacy
+components, but has been renamed since components no longer have an `element`
+property.
+
+**`didUpdate`** runs after the component has been updated, including all changes
+to `args` and any other changes which trigger a rerender of the component (i.e.
+computed properties being invalidated). It is effectively the same as
+`didUpdate` in legacy components, and can be used for update code.
+
+**`destroy`** is when the component is being destroyed, and can be used for
+cleanup code.
+
+Between these hooks and tracked/computed properties, users have replacements for
+all of the functionality provided by the legacy component hooks, with as little
+an API surface as possible to learn and remember.
+
+### Element Modifiers
+
+Glimmer components have outer HTML semantics, which means that whatever is put
+in the template is what we see in the final render. This also means that they
+can have _multiple_ root elements:
+
+```hbs
+<div ...attributes>
+  First root element
+</div>
+<div>
+  Second root element
+</div>
+```
+
+It is also possible that there are _no_ root elements:
+
+```hbs
+This is a text only component: {{@foo}}
+```
+
+This makes the `element` property of legacy components awkward, since there is
+no one single reference-able element. A previous proposal for solving this
+problem was a `bounds` property which contained `firstNode` and `lastNode`, but
+this could be very confusing to users since these nodes would likely have to be
+text nodes (how Glimmer tracks component boundaries internally), and would force
+users to write unintuitive looping logic to traverse nodes:
+
+```js
+let { firstNode, lastNode } = bounds;
+let currentNode = firstNode;
+
+while (currentNode !== lastNode) {
+  // do something with current node
+  currentNode = currentNode.nextSibling;
+}
+```
+
+This exposes users to low level APIs which may be useful to some, but in general
+are too complicated and nuanced. Instead, we can use new element modifiers which
+allow users to run code declaratively on the element:
+
+```hbs
+<div {{on-insert this.didInsertElement}} ...attributes>
+  First root element
+</div>
+<div {{on-insert this.didInsertElement}}>
+  Second root element
+</div>
+```
+
+This allows users to target which elements they care about directly. It also
+allows for code to be run on elements which are contained in conditionals,
+meaning users do not need to add messy logic to track component state
+themselves:
+
+```hbs
+{{#if shouldRender}}
+  <div
+    {{on-insert this.didInsertElement}}
+    {{on-destroy this.willDestroyElement}}
+  >
+
+  </div>
+{{/if}}
+```
+
+Finally, it means that users can bypass component code altogether by using
+helpers:
+
+```hbs
+<div {{on-insert (setScrollPosition @savedScrollPosition)}}></div>
+```
+
+This RFC proposes adding two element modifiers:
+
+1. `{{on-insert}}`, which runs immediately after the element is inserted.
```

Comment:
Is it fair to say, `on-insert` runs top-down?

Reply:
I believe this is bottom up? All child `{{on-insert}}`s will run before the parent does